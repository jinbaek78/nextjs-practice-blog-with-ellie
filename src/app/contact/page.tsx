import EmailForm from '@/components/EmailForm';
import { BsGithub, BsLinkedin, BsYoutube } from 'react-icons/bs';
export default function ContactPage() {
  const ICON_CLASS = 'hover:text-yellow-500';
  return (
    <article className="w-full flex flex-col items-center justify-center">
      <section className="">
        <div className="flex flex-col items-center justify-center">
          <h1 className="my-3 text-4xl font-bold">Contact Me</h1>
          <p className="text-lg font-medium">info@dream-coding.com</p>
          <div className="my-3 w-full flex justify-around font-bold text-5xl">
            <BsGithub className={ICON_CLASS} />
            <BsLinkedin className={ICON_CLASS} />
            <BsYoutube className={ICON_CLASS} />
          </div>
        </div>
      </section>
      <h2 className="my-5 text-4xl font-bold">Or Send me an email</h2>
      <EmailForm />
    </article>
  );
}
