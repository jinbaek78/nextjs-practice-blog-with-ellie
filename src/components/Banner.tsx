export type BannerData = {
  message: string;
  state: 'success' | 'error';
};

type Props = {
  banner: BannerData;
};
export default function Banner({ banner: { message, state } }: Props) {
  const isSuccess = state === 'success';
  const icon = isSuccess ? 'âœ…' : 'ðŸ”¥';

  return (
    <p
      className={`p-2 m-4 rounded-xl w-full text-center ${
        isSuccess ? 'bg-green-300' : 'bg-red-300'
      }`}
    >{`${icon} ${message}`}</p>
  );
}
