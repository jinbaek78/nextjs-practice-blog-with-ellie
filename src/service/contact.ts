import { EmailData } from './email';

export async function sendContactEmail(email: EmailData) {
  const response = await fetch('http://localhost:3000/api/contact', {
    method: 'POST',
    body: JSON.stringify(email),
    headers: {
      'Content-Type': 'application/json',
    },
  });

  const data = await response.json();
  if (!response.ok) {
    throw new Error(data.message || 'failed to request to server');
  }

  return data;
}
