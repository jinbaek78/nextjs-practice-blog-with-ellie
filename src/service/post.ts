import fs from 'fs/promises';
import path from 'path';
import { cwd } from 'process';

export type Post = {
  title: string;
  description: string;
  date: string;
  category: string;
  path: string;
  featured: boolean;
};

export async function getPostsData(): Promise<Post[]> {
  const postsData = await JSON.parse(
    await fs.readFile(`${path.join(cwd(), 'public', 'data', 'posts.json')}`, {
      encoding: 'utf-8',
    })
  );

  return postsData;
}
