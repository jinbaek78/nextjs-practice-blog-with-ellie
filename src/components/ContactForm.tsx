'use client';

import { ChangeEvent, FormEvent, useState } from 'react';
import Banner, { BannerData } from './Banner';

type Form = {
  from: string;
  subject: string;
  message: string;
};

export default function ContactForm() {
  const [form, setForm] = useState<Form>({
    from: '',
    subject: '',
    message: '',
  });

  const [banner, setBanner] = useState<BannerData | null>(null);

  const onChange = (e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    const { name, value } = e.target;
    setForm((prev) => ({ ...prev, [name]: value }));
  };

  const onSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    console.log(form);
    setBanner({ message: 'Success!', state: 'success' });
    // setTimeout(() => {
    //   setBanner(null);
    // }, 3000);
  };

  return (
    <section className="w-full max-w-md ">
      {banner && <Banner banner={banner} />}
      <form
        className="w-full  flex flex-col gap-2 m-4 p-4 bg-slate-700 rounded-xl text-white"
        onSubmit={onSubmit}
      >
        <label className="font-semibold" htmlFor="from">
          Your Email
        </label>
        <input
          className="text-black"
          type="email"
          id="from"
          name="from"
          required
          autoFocus
          value={form.from}
          onChange={onChange}
        />

        <label className="font-semibold" htmlFor="subject">
          Subject
        </label>
        <input
          className="text-black"
          type="text"
          id="subject"
          name="subject"
          required
          value={form.subject}
          onChange={onChange}
        />

        <label className="font-semibold" htmlFor="message">
          Message
        </label>
        <textarea
          className="text-black"
          rows={10}
          id="message"
          name="message"
          required
          value={form.message}
          onChange={onChange}
        />
        <button
          className="bg-yellow-300 text-black font-bold hover:bg-yellow-400"
          type="submit"
        >
          Submit
        </button>
      </form>
    </section>
  );
}
