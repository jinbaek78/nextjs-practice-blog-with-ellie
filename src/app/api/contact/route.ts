import { sendEmail } from '@/service/email';
import { NextResponse } from 'next/server';
import * as yup from 'yup';

const bodySchema = yup.object().shape({
  from: yup.string().email().required(),
  subject: yup.string().required(),
  message: yup.string().min(3).required(),
});

export async function POST(req: Request) {
  const body = await req.json();

  if (!bodySchema.isValidSync(body)) {
    return NextResponse.json(
      { message: 'it is invalid format' },
      { status: 400 }
    );
  }

  return sendEmail(body) //
    .then(() =>
      NextResponse.json(
        { message: 'your email has successfully delivered' },
        { status: 200 }
      )
    )
    .catch((error) => {
      console.error(error);
      return NextResponse.json(
        {
          message: 'your delivery has failed, please, try again',
        },
        { status: 500 }
      );
    });
}
