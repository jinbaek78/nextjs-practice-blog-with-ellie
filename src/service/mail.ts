import { BannerData } from '@/components/Banner';
import { Form } from '@/components/ContactForm';
import { object, string, number, date, InferType } from 'yup';

export async function sendEmail(
  data: Form
  // callback: (banner: BannerData) => void
): Promise<BannerData> {
  let formSchema = object({
    subject: string().required(),
    from: string().email().required(),
    message: string().min(2).required(),
  });

  let dataValidationResult;
  try {
    dataValidationResult = await formSchema.validate(data);
  } catch (err) {
    console.log(err);
    return { message: 'please, write the right format', state: 'error' };
  }

  const formHeader = new Headers();
  formHeader.append('Content-Type', 'application/json');

  const requestOptions = {
    method: 'POST',
    headers: formHeader,
    body: JSON.stringify(data),
  };

  let res;
  let resultBanner: BannerData;
  try {
    res = await fetch('http://localhost:3000/api/contact', requestOptions);
  } catch (err) {
    console.log(err);
    return (resultBanner = { message: 'fail', state: 'error' });
  }

  return res.status === 200
    ? { message: 'success', state: 'success' }
    : { message: 'fail', state: 'error' };

  // return fetch('http://localhost:3000/api/contact', requestOptions)
  //   .then((response) => response.json())
  //   .then((result) => console.log(result))
  //   .catch((error) => console.log('error', error));
}
