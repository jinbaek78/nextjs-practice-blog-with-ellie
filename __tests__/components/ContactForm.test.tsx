import {
  render,
  screen,
  fireEvent,
  waitFor,
  queryByDisplayValue,
  getByTestId,
  act,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import ContactForm from '@/components/ContactForm';
import { sendContactEmail } from '@/service/contact';
import Banner, { BannerData } from '@/components/Banner';

jest.mock('@/service/contact');
jest.mock('@/components/Banner');
// jest.mock('@/components/Banner', () => {
//   const originalModule = jest.requireActual('@/components/Banner');
//   return {
//     __esModule: true,
//     ...originalModule,
//     Banner: jest.fn(async () => ({ message: 'success', state: 'success' })),
//   };
// });

describe('ContactForm', () => {
  afterEach(() => {
    (
      sendContactEmail as jest.MockedFunction<typeof sendContactEmail>
    ).mockReset();
    // (Banner as jest.MockedFunction<typeof Banner>).mockReset();
  });

  xit('renders correctly', () => {
    const { container } = render(<ContactForm />);
    expect(container).toMatchSnapshot();
  });

  xit('renders basic ui correctly', () => {
    render(<ContactForm />);

    expect(Banner as jest.MockedFunction<typeof Banner>).not.toBeCalled();
    expect(screen.getByText('Your Email')).toBeInTheDocument();
    expect(screen.getByText('Subject')).toBeInTheDocument();
    expect(screen.getByText('Message')).toBeInTheDocument();
    expect(screen.getByText('Submit')).toBeInTheDocument();
  });

  it('displays a success message banner when we receive a success message and automatically hid it after 3 seconds', async () => {
    const successBanner: BannerData = { message: 'success', state: 'success' };
    (sendContactEmail as jest.Mock).mockImplementation(async (form) => {
      return successBanner;
    });
    // mocked Banner
    (Banner as jest.MockedFunction<typeof Banner>).mockImplementation(
      ({ banner: { message, state } }) => {
        return (
          <p>
            {state === 'success' ? 'âœ…' : 'ðŸ”¥'} {message}
          </p>
        );
      }
    );
    render(<ContactForm />);
    // expect(Banner as jest.MockedFunction<typeof Banner>).toBeCalled();
    const fromInput = screen.getByTestId('from');

    await fireEvent.change(fromInput, {
      target: { value: 'jin123@google.com' },
    });

    await waitFor(() =>
      expect(screen.getByTestId('from')).toHaveDisplayValue('jin123@google.com')
    );
    const subjectInput = screen.getByTestId('subject');
    fireEvent.change(subjectInput, {
      target: { value: 'hi' },
    });
    await waitFor(() =>
      expect(screen.getByTestId('subject')).toHaveDisplayValue('hi')
    );

    const messageInput = screen.getByTestId('message');

    //
    await fireEvent.change(messageInput, {
      target: { value: 'hi, there' },
    });
    await waitFor(() =>
      expect(screen.getByTestId('message')).toHaveDisplayValue('hi, there')
    );

    //
    const submitButton = screen.getByRole('button');

    await fireEvent.click(submitButton);

    await waitFor(() => {
      expect(screen.getByText('âœ… success')).toBeInTheDocument();
    });
  });

  it('displays a error message banner when we receive a error message and automatically hid it after 3 seconds', async () => {
    const errorBanner: BannerData = { message: 'error', state: 'error' };
    (sendContactEmail as jest.Mock).mockImplementation(async (form) => {
      throw new Error(errorBanner.message);
    });
    (Banner as jest.MockedFunction<typeof Banner>).mockImplementation(
      ({ banner: { message, state } }) => {
        console.log('got state:', state);
        return (
          <p>
            {state === 'success' ? 'âœ…' : 'ðŸ”¥'} {message}
          </p>
        );
      }
    );
    render(<ContactForm />);
    const fromInput = screen.getByTestId('from');

    await fireEvent.change(fromInput, {
      target: { value: 'jin123@google.com' },
    });

    await waitFor(() =>
      expect(screen.getByTestId('from')).toHaveDisplayValue('jin123@google.com')
    );
    const subjectInput = screen.getByTestId('subject');
    fireEvent.change(subjectInput, {
      target: { value: 'hi' },
    });
    await waitFor(() =>
      expect(screen.getByTestId('subject')).toHaveDisplayValue('hi')
    );

    const messageInput = screen.getByTestId('message');

    //
    await fireEvent.change(messageInput, {
      target: { value: 'hi, there' },
    });
    await waitFor(() =>
      expect(screen.getByTestId('message')).toHaveDisplayValue('hi, there')
    );

    //
    const submitButton = screen.getByRole('button');

    await fireEvent.click(submitButton);

    await waitFor(() => {
      expect(screen.getByText('ðŸ”¥ error')).toBeInTheDocument();
    });
    await waitForElementToBeRemoved(screen.getByText('ðŸ”¥ error'), {
      timeout: 3000,
    });
  });
});
