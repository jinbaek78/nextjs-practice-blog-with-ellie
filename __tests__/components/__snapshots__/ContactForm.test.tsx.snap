// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm renders correctly 1`] = `
<div>
  <section
    class="w-full max-w-md "
  >
    <form
      class="w-full  flex flex-col gap-2 m-4 p-4 bg-slate-700 rounded-xl text-white"
    >
      <label
        class="font-semibold"
        for="from"
      >
        Your Email
      </label>
      <input
        class="text-black"
        data-testid="from"
        id="from"
        name="from"
        required=""
        type="email"
        value=""
      />
      <label
        class="font-semibold"
        for="subject"
      >
        Subject
      </label>
      <input
        class="text-black"
        data-testid="subject"
        id="subject"
        name="subject"
        required=""
        type="text"
        value=""
      />
      <label
        class="font-semibold"
        for="message"
      >
        Message
      </label>
      <textarea
        class="text-black"
        data-testid="message"
        id="message"
        name="message"
        required=""
        rows="10"
      />
      <button
        class="bg-yellow-300 text-black font-bold hover:bg-yellow-400"
        type="submit"
      >
        Submit
      </button>
    </form>
  </section>
</div>
`;
